sanitation of Infer.

paulp opened the pull request:
Merge frustration led me to stick to one file for a while.

paulp commented:
help indeed. Guess one has to get the command line args in juuuust the right order.

paulp commented:
Nobody sane will pursue fine-grained commits if it means the tests will never pass for them... here's one of the random inexplicable failures. The commit being tested is the removal of comments. I am not sure what percentage of our failures are meaningless, but I do know I get an unreasonable amount of mail about jenkins failures.

```
test.interactive:
  [partest] Running 20 tests in 'presentation' at Sun May 19 09:21:04 CEST 2013
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0x7) at pc=0x00007fbc2e812941, pid=3925, tid=140446042928896
#
# JRE version: 6.0_33-b04
# Java VM: Java HotSpot(TM) 64-Bit Server VM (20.8-b03 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.so+0x277941]  Copy::pd_disjoint_words(HeapWord*, HeapWord*, unsigned long)+0x41
#
# An error report file with more information is saved as:
# /localhome/jenkins/a/workspace/pr-checkin-per-commit/hs_err_pid3925.log
#
# If you would like to submit a bug report, please visit:
#   http://java.sun.com/webapps/bugreport/crash.jsp
#
/localhome/jenkins/a/workspace/pr-checkin-per-commit/jenkins-scripts/build: line 33:  3925 Aborted                 ant $antArgs $scalacArgs nightly
+ buildState=134
```

retronym commented:
Something has made `presentation/implicit-member` non-deterministic of late; Miguel noticed the same failure on his PR.

As for JVM crashes, we should make Jenkins archive `hs_err*.pid` files as build artifacts. Otherwise we're left guessing as to the root cause. @lrytz: is there a way we can do that across the board? I guess we'd need to edit the likes of https://github.com/lrytz/jenkins-config/blob/master/jobs/pr-checkin-per-commit/config.xml#L99.

paulp commented:
There are even more ambitious booches in this round.

```
  [partest] ok  333 - run/macro-declared-in-method            
  [partest] ok  334 - run/sort.scala                          
  [partest] ok  335 - run/t5276_1a.scala                      
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0x7) at pc=0x00007fe4d903dff9, pid=7583, tid=140620655421184
#
# JRE version: 6.0_33-b04
# Java VM: Java HotSpot(TM) 64-Bit Server VM (20.8-b03 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.so+0x388ff9]  [partest] ok  336 - run/t3984.scala                         
[thread 140620666885888 also had an error]
  CollectedHeap::fill_with_object(HeapWord*, unsigned long, bool)+0xb9
#
# An error report file with more information is saved as:
# /localhome/jenkins/a/workspace/pr-rangepos-per-commit/hs_err_pid7583.log
# [ timer expired, abort... ]
./jenkins-scripts/job/scala-checkin-rangepos: line 19:  7583 Aborted                 ant -Dscalac.args.optimise=-optimise $scalacArgs test.suite test.continuations.suite test.scaladoc
FATAL: Unable to delete script file /tmp/hudson4737021428598931227.sh
hudson.util.IOException2: remote file operation failed: /tmp/hudson4737021428598931227.sh at hudson.remoting.Channel@6b2fd0b3:lampscalatst2-a
    at hudson.FilePath.act(FilePath.java:900)
    at hudson.FilePath.act(FilePath.java:877)
```

paulp commented:
NOLITTER!

paulp commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 18115916)
:cat: Roger! Rebuilding pr-rangepos-per-commit, pr-checkin-per-commit for 34ab4b69, b60147fc, 918d98d9, 62a7e796, 48fe4ad6, 1591b09d, a582a876, 10fce94d, c31facb8, 84181eec, 7d687b85, 920f6b38, d825b644, c643cc65, d20fc122, 6df2ccb4. :rotating_light:

scala-jenkins commented:
(kitty-note-to-self: ignore 18115914)
:cat: cleaning up... sorry! :cat:

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -170,46 +174,34 @@ trait Infer extends Checkable {
+      case TypeVar(_, constr)                           =>
+        excludedVars += tv
+        try apply(constr.inst)
+        finally excludedVars -= tv
```
This match is driven entirely with guards; as such I find the `if/else` formulation simpler.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -170,46 +174,34 @@ trait Infer extends Checkable {
+    case TypeRef(pre, _, args)                          => isFullyDefined(pre) && (args forall isFullyDefined)
+    case SingleType(pre, _)                             => isFullyDefined(pre)
+    case RefinedType(ts, _)                             => ts forall isFullyDefined
+    case TypeVar(_, constr) if constr.inst == NoType    => false
```
`if !constr.instValid` ?

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -170,46 +174,34 @@ trait Infer extends Checkable {
     }
   }
 
+  @inline final def falseIfNoInstance(body: => Boolean): Boolean =
```
Seems like this could be private.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -309,102 +291,87 @@ trait Infer extends Checkable {
+      if (settings.debug) {
+        Console.println(context)
+        Console.println(tree)
+        Console.println("" + pre + " " + sym.owner + " " + context.owner + " " + context.outer.enclClass.owner + " " + sym.owner.thisType + (pre =:= sym.owner.thisType))
```
While we're here:

``` scala
logDebug(sm"""$context
             |$tree
             |$pre ${sym.owner} ${context.owner} ${context.outer.enclClass.owner} ${sym.owner.thisType}  ${pre =:= sym.owner.thisType}""")
```

Or something a touch more self-explanatory

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -461,50 +422,50 @@ trait Infer extends Checkable {
+      if (conforms)
+        try solve() catch { case _: NoInstance => null }
+      else
+        null
```
Do you think we ended up with `null` here rather than `Option` or our own result ADT on grounds of performance? Or maybe just through path-of-least-resistance code evolution? If it is the former, or that's our best suspicion, we ought to document that here. Otherwise, we could do away with the `null`.

paulp replied:
Pure path of least resistance. I was all over this one, just had to commit something lest it be another lost-in-the-weeds adventure. (You have no idea how much code like this has never been checked in because I went for one more thing...) Anyway, it's definitely a thing to get.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -515,22 +476,20 @@ trait Infer extends Checkable {
         def setInst(tp: Type): Type = {
           tvar setInst tp
-          assertNonCyclic(tvar)//debug
+          assert(tvar.constr.inst != tvar, tvar.origin)
```
``` sm
assert(tvar.constr.inst != tvar, s"cylic TypeVar ${tvar.origin}")
```

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -834,9 +792,50 @@ trait Infer extends Checkable {
+        case (_, pos, _) if !allArgsArePositional(pos) && !sameLength(formals, mt.params) => false // different length lists and all args not positional
+        case (args, pos, _)                                                               => typesCompatible(reorderArgs(args, pos))
+      }
+      compareLengths(argtpes0, formals) match {
```
This calls out for a version of `compareLengths` that returns an ADT `LT`, `EQ`, `GT`. `compareLengths` and `hasLength` could also move to `Collections`.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -922,105 +861,76 @@ trait Infer extends Checkable {
+        val silent = context.makeSilent(reportAmbiguousErrors = false)
+        val result = newTyper(silent).infer.isApplicable(undetparams, ftpe, argtpes0, pt)
+        if (silent.hasErrors && !pt.isWildcard)
+          applicableExpectingPt(WildcardType) // second try
```
Much better. I was to blame in 510ebeca9, where I got rid of the duplication in calls to `isApplicable`, but missed this much simpler recursive formulation.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -1033,90 +943,34 @@ trait Infer extends Checkable {
+          case _                    => false
+        }
+      case _ => false
     }
```
When @adriaanm's tuple matching optimization lands, or sooner if this method isn't called much:

``` scala
(ftpe1, ftpe2) match {
  case (MethodType(_, rtpe1), MethodType(_, rtpe2)) =>
    rtpe1 <:< rtpe2 || rtpe2.typeSymbol == ObjectClass
  case _ => false
}
```

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -1500,38 +1352,24 @@ trait Infer extends Checkable {
-      case OverloadedType(pre, alts) => tryTwice { isSecondTry =>
-        val alts0          = alts filter (alt => isWeaklyCompatible(pre.memberType(alt), pt))
+    def inferExprAlternative(tree: Tree, pt: Type): Tree = {
+      def tryOurBests(pre: Type, alts: List[Symbol], isSecondTry: Boolean): Unit = {
```
This method name has a pleasing ring of Baden Powell to it.

retronym commented:
From @adriaanm's [release guide](https://docs.google.com/a/typesafe.com/document/d/10BiyBbSDbwb2tJCSJcV5pAThyeiu_uiObaBIG_MZBKo/edit#):

> are all jenkins nodes up? reboot them perhaps for good measure (sometimes jenkins fails spuriously due to low swap space)

Wild guess, but perhaps that JVM crash might be the symptom.

retronym commented:
LGTM. Doubleplusgood, in fact. Feel free to defer any or all of my minor comments to a followup.

paulp commented:
Re presentation/implicit-member, #2566.

paulp committed:
Cleaning up isApplicable and neighbors.

I'd really like these methods to receive greater scrutiny.
To that end I'm scrutinizing them.

paulp committed:
Cleaning crew moves down the hall.

Swabbing the decks of isAsSpecific and isApplicableSafe.
Wow isApplicableSafe, you don't really need to declare and
allocate a two-argument case class in order to manage a
second method call.

paulp committed:
Commented out code from 2006.

Yes, 2006. Perhaps we should have a "sunset provision"
which removes commented-out code after seven years. Or even
slightly sooner.

paulp committed:
Moving on to checkBounds.

So much dust and debris, so little time.

paulp committed:
Commented out code from 2007.

It's not exactly 2006-era code but it's still a little ripe.

paulp committed:
Scrubbing up inferPolyAlternatives.

Now you can take inferPolyAlternatives home to meet your mother.

paulp committed:
Help makeFullyDefined allocate fewer ListBuffers.

paulp committed:
Splashed soap and water on checkAccessible.

Yes checkAccessible, you patiently waited your turn.
Do you like your new clothes? Oh, still not clean enough?
I understand, but we must leave time for the other methods.
Now run along and check something.

paulp committed:
Remove isPossiblyMissingArgs.

It has been hardcoded to return false for a long while anyway.

paulp committed:
Make cookJavaRawInfo return this.type.

A lot more useful than Unit.

paulp committed:
Wipe some dirt off protoTypeArgs's face.

He's a good kid, even if not the most hygenic.

paulp committed:
Tried to follow own advice with isCoercible.

Only to discover that it's really hard to move isCoercible
anywhere because it wants to call inferView which, despite its
suggestive name, is not visible in Infer. So I did what I
could and documented it a little.

paulp committed:
Commented out code from 2008.

paulp committed:
Annotated return type of a method.

Because nobody should have to guess a return type like
(List[Type], Array[Int], Boolean). In fact nobody should
ever experience such a return type at all.

paulp committed:
Pull some big chunks out of checkAccessible.

To streamline the essential logic.

paulp committed:
Swabbing exprTypeArgs.

Oh exprTypeArgs, you will send us to the poorhouse with
your spendy ways. So many tuples have been dying needlessly.

paulp commented:
PLS SYNCH

scala-jenkins commented:
(kitty-note-to-self: ignore 18123524)
:cat: Synchronaising! :pray:

retronym commented:
Pleading for Jenkins help here: https://groups.google.com/d/msg/scala-internals/azlMUvABpME/uxFG1OeKOgEJ

retronym commented:
One qualification to my review: reading the commits in sequence makes it pretty clear that these are all straight up refactorings. But if you hit one in isolation, you have to use a little inference of your own to know that "Moving on to checkBounds. So much dust and debris, so little time.", etc, means as much. I'd infer as much based on the author and the fact no tests were updated. But it doesn't hurt to add an addendum to these sort of commits to make that explicit.

paulp commented:
Yeah... I ran out of ways to say "I made this code more like it should have been in the first place" a few years back and there's no telling what I'll come up with at this point. Maybe I just need some kind of template so I don't have to try to come up with original prose.

retronym commented:
PLS REBUILD/pr-rangepos-per-commit@661414e
PLS REBUILD/pr-checkin-per-commit@96f1993
PLS REBUILD/pr-rangepos-per-commit@96f1993
PLS REBUILD/pr-checkin-per-commit@e058a80
PLS REBUILD/pr-rangepos-per-commit@e4da021
PLS REBUILD/pr-checkin-per-commit@e4da021
PLS REBUILD/pr-rangepos-per-commit@a2bafaa
PLS REBUILD/pr-checkin-per-commit@2cdcccd

scala-jenkins commented:
(kitty-note-to-self: ignore 18136852)
:cat: Roger! Rebuilding pr-checkin-per-commit for 2cdcccd6. :rotating_light:

paulp commented:
Is this related to any open or recently merged plugin fix, like #2562 ?

```
docs.continuations-plugin:
    [mkdir] Created dir: /localhome/jenkins/a/workspace/pr-checkin-per-commit/build/scaladoc/continuations-plugin
 [scaladoc] Element '/localhome/jenkins/a/workspace/pr-checkin-per-commit/src/continuations/plugin' does not exist.
 [scaladoc] Element '/localhome/jenkins/a/workspace/pr-checkin-per-commit/src/continuations/plugin' does not exist.
```

retronym commented:
That error predated the merge. I think it's unrelated.

paulp commented:
Is it all some kind of crazy dream? Given the unlikelihood of ever seeing all green checkmarks again, I'm going to take the offer to defer some fiddling to a followup.
